packages/apex-node/src/coverageReporters/coverageReporter.ts

```diff
@@ -0,0 +1,163 @@
+import * as reports from 'istanbul-reports';
+import * as libCoverage from 'istanbul-lib-coverage';
+import * as path from 'path';
+import { glob } from 'glob';
+import * as fs from 'fs';
+
+const startOfSource = (source: string): number => {
+  if (source) {
+    return source.search(/\S/) || 0;
+  }
+  return 0;
+};
+const endOfSource = (source: string): number => {
+  if (source) {
+    return source.search(/\S$/) || 0;
+  }
+  return 0;
+};
+
+export type CoverageReportFormats = reports.ReportType;
+
+export const DefaultWatermarks: libReport.Watermarks = {
+  statements: [50, 75],
+  functions: [50, 75],
+  branches: [50, 75],
+  lines: [50, 75]
+};
+
+export const DefaultReportOptions: Partial<reports.ReportOptions> = {
+  clover: { file: 'clover.xml', projectRoot: '.' },
+  cobertura: { file: 'cobertura.xml', projectRoot: '.' },
+  'html-spa': {
+    verbose: false,
+    skipEmpty: false,
+    subdir: 'html-spa',
+    linkMapper: undefined,
+    metricsToShow: ['lines', 'statements', 'branches']
+  },
+  html: {
+    verbose: false,
+    skipEmpty: false,
+    subdir: 'html',
+    linkMapper: undefined
+  },
+  json: { file: 'coverage.json' },
+  'json-summary': { file: 'coverage-summary.json' },
+  lcovonly: { file: 'lcovonly.info', projectRoot: '.' },
+  none: {} as never,
+  teamcity: { file: 'teamcity.txt', blockName: 'coverage' },
+  text: { file: 'text.txt', maxCols: 160, skipEmpty: false, skipFull: false },
+  'text-summary': { file: 'text-summary.txt' }
+};
+
+export interface CoverageReporterOptions {
+  reportFormats?: CoverageReportFormats[];
+  reportOptions?: Partial<typeof DefaultReportOptions>;
+  watermark?: typeof DefaultWatermarks;
+}
+
+export class CoverageReporter {
+  private readonly coverageMap: libCoverage.CoverageMap;
+  constructor(
+    private readonly coverage: ApexCodeCoverageAggregate | ApexCodeCoverage,
+    private readonly reportDir: string,
+    private readonly sourceDir: string,
+    private readonly options?: CoverageReporterOptions
+  ) {
+    this.coverageMap = this.buildCoverageMap();
+  }
+
+  public generateReports(): void {
+    const context = libReport.createContext({
+      dir: this.reportDir,
+      defaultSummarizer: 'nested',
+      watermarks: this.options?.watermark || DefaultWatermarks,
+      coverageMap: this.coverageMap
+    });
+    const formats = this.options?.reportFormats || ['text-summary'];
+    formats.forEach(format => {
+      const report = reports.create(
+        format,
+        this.options?.reportOptions[format] || DefaultReportOptions[format]
+      );
+      report.execute(context);
+    });
+  }
+
+  private buildCoverageMap(): libCoverage.CoverageMap {
+    const coverageMap = libCoverage.createCoverageMap();
+    this.coverage.records.forEach(
+      (record: ApexCodeCoverageRecord | ApexCodeCoverageAggregateRecord) => {
+        const fileCoverageData: libCoverage.FileCoverageData = {} as libCoverage.FileCoverageData;
+        fileCoverageData.fnMap = {};
```

Comment:
There's several empty arrays being set here. Is this something that Istanbul needs, or can we remove these lines?

Reply:
I will have a look at removing the empty elements.

Reply:
Randi, the empty elements are required by istanbul.